source/extensions/filters/common/expr/custom_cel/custom_cel_vocabulary.h

```diff
@@ -0,0 +1,163 @@
+  // (2) registerFunctions will add it to the CEL function registry.
+  // The function mapping for a stateful/lazy function which is added to the
+  // activation in fillActivation must be compatible with the function descriptor
+  // added to the registry in registerFunctions.
+  // Stateless/static functions do not need to be registered with the activation,
+  // only with the CEL function registry.
+  //
+  // The activation contains two sets of vocabulary (variables and functions):
+  // (1) the Envoy defined CEL vocabulary (Request, Response, Connection, etc.)
+  // which is added to the activation in the evaluator's createActivation function.
+  // (2) the custom CEL vocabulary, which is added to the activation in fillActivation.
+  // The Envoy defined vocabulary is registered first and the custom CEL vocabulary second.
+  //
+  // The implementer of this interface can choose to override the Envoy defined CEL vocabulary
+  // (Request, Response, Connection, etc.) with a custom implementation.
+  // The mappings for the Envoy defined CEL vocabulary would have to be removed from the
+  // activation and replaced with the mappings for the custom CEL vocabulary.
+  //
+  // It is possible to remove entries from an activation.
+  // It is not possible to remove entries from the CEL function registry.
+  virtual void fillActivation(google::api::expr::runtime::Activation* activation,
+                              Protobuf::Arena& arena, const StreamInfo::StreamInfo& info,
+                              const Http::RequestHeaderMap* request_headers,
+                              const Http::ResponseHeaderMap* response_headers,
+                              const Http::ResponseTrailerMap* response_trailers) PURE;
+
+  // registerFunctions:
+  //
+  // Registers both stateful/lazy and stateless/static functions in the CEL function registry.
+  // The registry, with its functions, is used by the CEL expression builder to
+  // build CEL expressions.
+  // The registry contains two groups of functions.
+  // (1) Built-in CEL  functions. ("+", "-" "*", "!", etc.)
+  // (2) Custom CEL functions.
+  // (There are currently no Envoy defined functions which are registered).
+  // The CelExpressionBuilder which is used to create CelExpressions contains a registry
+  // in which the built-in CEL functions have been registered.
+  // The custom CEL functions are registered in the registry using registerFunctions.
+  // Stateful/lazy functions must also be added to the activation using fillActivation.
+  // Stateless/static functions do not have to be added to the activation.
+  // The stateful/lazy function activation mapping must match its counterpart
+  // stateful/lazy function descriptor in the registry.
+  // Once registered, the function registration cannot be removed from the registry
+  // or overridden, as it can for an activation mapping.
+  virtual void registerFunctions(google::api::expr::runtime::CelFunctionRegistry* registry) PURE;
```

Comment:
Is there a difference in lifecycle invocations between `fillActivation` and `registerFunctions`? I.e. `registerFunctions` is called only in the factory constructor for the extension, whereas `fillActivation` must occur on every request?

Reply:
The rbac filter rules engine has set up so that there is a difference.  

There's a cel expression builder with cel function registry that's created for a single RBAC rule with its multiple policies.  
`registerFunctions` modifies this cel function registry.  

On the other hand, an activation is created for every individual policy.  `fillActivation` modifies this activation.

Reply:
I see. I think it's worth providing this level of detail, since you already have really great explanations in this file and this is interesting detail in that context.